Sources/Common/Background Queue/QueueRunRequest.swift

```diff
@@ -38,106 +41,124 @@ public class CioQueueRunRequest: QueueRunRequest {
     }
 
     private func startNewRequestRun() {
-        runTasks(lastRanTask: nil)
+        threadUtil.runBackground {
+            self.runTasks()
+        }
     }
 
     // Disable swiftlint because function at this time isn't too complex to need to make it smaller.
     // Many of the lines of this function are logging related.
     // swiftlint:disable:next function_body_length
-    private func runTasks(
-        lastRanTask: QueueTaskMetadata?,
-        lastFailedTask: QueueTaskMetadata? = nil
-    ) {
-        // get the inventory before running each task. If a task was added to the queue while the last task was being
-        // executed, we can assert that new task will execute during this run.
-        let queueInventory = storage.getInventory()
-
-        guard let nextTaskToRunInventoryItem = queryRunner.getNextTask(
-            queueInventory,
-            lastRanTask: lastRanTask,
-            lastFailedTask: lastFailedTask
-        )
-        else {
-            // we hit the end of the current inventory. Done!
-            return doneRunning()
-        }
+    internal func runTasks() {
+        // Variables that power the logic of the queue run loop
+        var lastRanTask: QueueTaskMetadata?
+        var lastFailedTask: QueueTaskMetadata?
+        var continueRunnning = true
+        let whileLoopWait = DispatchGroup() // make async operations perform synchronously in while loop
```

Comment:
is there any specific reason we opted for `DispatchGroup` rather `Semaphores` I believe they do serve somewhat a similar purpose?

Reply:
Using a semaphore didn't come to mind. In fact, I have never used a semaphore in iOS before. 

`DispatchGroup` is popular in iOS/mac development and will more then likely see the class used. I am not sure about the history of `DispatchGroup`, but I would guess that it's an abstraction that is provided by Apple to provide a more convenient API during development. 

The same could be said about using `DispatchQueue` for creating and executing code on a thread. Or, using `Core Data` for writing to a database. All are abstraction APIs provided by Apple that are popular to use in the development community. However, there are usually alternatives to these abstractions. 

Long story short, `DispatchGroup` is popular, iOS devs are familiar with it, and it's simple to use.